### PR TITLE
feat(suite-common): Transactions

### PR DESCRIPTION
**Pull request by juriczech**
## Description
Moving transactions reducer and actions from suite package to `/suite-common/wallet-transaction`. This will be new package, that will allow for shared logic between different applications (mobile, desktop, etc). 

Refactor from "vanilla" redux to [ReduxToolkit](https://redux-toolkit.js.org/) has been done as discussed on latest code improvements. 

Only one thunk remained in /suite package because it uses browser API for saving file that won't be working with mobile so this is yet to be refactored as a separate issue later on. 

## Related Issue
https://github.com/trezor/trezor-suite/issues/5709

### CODE REVIEW

File: packages/suite/src/support/extraDependencies.ts [hunk 1]
```diff
@@ -28,5 +29,15 @@ export const extraDependencies: ExtraDependencies = {
                 state[backend.key].backends = backend.value;
             });
         },
+        storageLoadTransactions: (state, { payload }: StorageLoadAction) => {
```
Comment (Nodonisko) [inline packages/suite/src/support/extraDependencies.ts]: Missing type for state

File: suite-common/wallet-transactions/src/slice.ts [hunk 2]
```diff
@@ -149,4 +156,4 @@ export const {
     replaceTransaction,
     resetTransaction,
 } = transactionSlice.actions;
-export const { reducer } = transactionSlice;
+export const reducer = transactionSlice.prepareReducer;
```
Comment (Nodonisko) [inline suite-common/wallet-transactions/src/slice.ts]: preparrTransactionsReducer

File: suite-common/wallet-transactions/src/transactionThunks.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            vout: [],
+            size: 0,
+            totalInput: '0',
+            totalOutput: '0',
+        },
+    };
+    dispatch(transactionActions.addTransaction({ transactions: [fakeTx], account }));
+});
+
+export const prepareFetchTransactionsThunk = ({
```
Comment (Nodonisko) [inline suite-common/wallet-transactions/src/transactionThunks.ts]: Use `createThunk` from redux-utils package everywhere in this file

Also you should be able to get rid of this "prepare" function because you have extra avaialable

File: suite-common/wallet-transactions/src/transactionThunks.ts [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        },
+        { state: any }
+    >(
+        `${modulePrefix}/addFakePendingTransaction`,
+        async (
+            { account, page, perPage, noLoading = false, recursive = false },
+            { dispatch, getState },
+        ) => {
+            if (!isTrezorConnectBackendType(account.backendType)) return; // skip unsupported backend type
+            const { transactions } = getState().wallet.transactions;
```
Comment (Nodonisko) [inline suite-common/wallet-transactions/src/transactionThunks.ts]: never access state directly, getState must be always used in selector

`const transaction = selectTransactions(getState())`

Please fix this everywhere in all thunks

File: suite-common/wallet-transactions/src/transactionReducer.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
+import { Account, WalletAccountTransaction } from '@suite-common/wallet-types';
+import { findTransaction } from '@suite-common/wallet-utils';
+import { settingsCommonConfig } from '@suite-common/suite-config';
+import { createReducerWithExtraDeps, matchesActionType } from '@suite-common/redux-utils';
+
+import { transactionActions } from './transactionActions';
+
+export interface TransactionSliceState {
```
Comment (Nodonisko) [inline suite-common/wallet-transactions/src/transactionReducer.ts]: TransactionState

File: suite-common/wallet-transactions/src/transactionReducer.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    transactions: { [key: string]: WalletAccountTransaction[] };
+}
+
+export const initialState: TransactionSliceState = {
+    isLoading: false,
+    error: null,
+    transactions: {},
+};
+
+interface TransactionSliceRootState {
```
Comment (Nodonisko) [inline suite-common/wallet-transactions/src/transactionReducer.ts]: TransactionRootState

File: suite-common/wallet-transactions/src/transactionReducer.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            matchesActionType(extra.actionTypes.fiatRateUpdate),
+            extra.reducers.txFiatRateUpdate,
+        )
+        .addMatcher(
+            matchesActionType(extra.actionTypes.removeAccount),
+            extra.reducers.removeAccountsTransactions,
+        );
+});
+
+export const selectIsLoading = (state: TransactionSliceRootState) =>
```
Comment (Nodonisko) [inline suite-common/wallet-transactions/src/transactionReducer.ts]: use better names because these selector are exported out of this file and in context of wallet isLoading can mean anything

`selectIsLoadingTransactions`
`selectTransactionsError`

etc.

File: suite-common/redux-utils/src/createReducerWithExtraDeps.ts [hunk 8]
```diff
@@ -7,19 +7,20 @@ type NotFunction<T> = T extends Function ? never : T;
...
+) => {
+    const prepareReducer = (extraDeps: ExtraDependencies) =>
         createReducer(initialState, builder =>
             builderCallback(builder, {
                 actionTypes: extraDeps.actionTypes,
                 actions: extraDeps.actions,
                 reducers: extraDeps.reducers,
             }),
         );
+    return { prepareReducer };
```
Comment (Nodonisko) [inline suite-common/redux-utils/src/createReducerWithExtraDeps.ts]: please revert this changes, this was mistake it should return prepare function directly without any wrapping into object

File: suite-common/redux-utils/src/extraDependenciesType.ts [hunk 9]
```diff
@@ -8,6 +8,9 @@ import { NotificationEventPayload } from '@suite-common/notifications';
 import { ActionType, SuiteCompatibleSelector, SuiteCompatibleThunk } from './types';
 
 type StorageLoadReducer = (state: any, action: { type: any; payload: any }) => void;
+type StorageLoadTransactionsReducer = (state: any, action: { type: any; payload: any }) => void;
```
Comment (Nodonisko) [inline suite-common/redux-utils/src/extraDependenciesType.ts:17]: return type shouldn't be void but never, that also wrong for `StorageLoadReducer`

File: suite-common/wallet-transactions/src/transactionReducer.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+                        (!existingTx.blockTime && transaction.blockTime)
+                    ) {
+                        // pending tx got confirmed (blockHeight changed from undefined/0 to a number > 0)
+                        accountTxs[existingTxIndex] = { ...transaction };
+                    }
+                }
+            });
+        })
+        .addMatcher(
+            matchesActionType(extra.actionTypes.storageLoad),
```
Comment (Nodonisko) [inline suite-common/wallet-transactions/src/transactionReducer.ts]: (action) => action.type === extra.actionTypes.storageLoad

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/wallet-transactions/src/transactionReducer.ts (hunk 7 above)
Comment (juriczech) [reply suite-common/wallet-transactions/src/transactionReducer.ts]: Yeah good point. I was thinking about it and only kept it like this because you always assign the selector value to another variable but it could be confusing when also importing other `isLoading` variable to same file. Renamed.

File: suite-common/wallet-transactions/src/transactionReducer.ts (hunk 10 above)
Comment (juriczech) [reply suite-common/wallet-transactions/src/transactionReducer.ts]: https://github.com/trezor/trezor-suite/pull/5981/commits/86973889f127ee510e9e5b3f9fd0d10ad515d1c8

File: suite-common/wallet-transactions/src/transactionReducer.ts (hunk 7 above)
Comment (juriczech) [reply suite-common/wallet-transactions/src/transactionReducer.ts]: https://github.com/trezor/trezor-suite/pull/5981/commits/86973889f127ee510e9e5b3f9fd0d10ad515d1c8

File: suite-common/wallet-transactions/src/transactionReducer.ts (hunk 6 above)
Comment (juriczech) [reply suite-common/wallet-transactions/src/transactionReducer.ts]: https://github.com/trezor/trezor-suite/pull/5981/commits/86973889f127ee510e9e5b3f9fd0d10ad515d1c8

File: suite-common/wallet-transactions/src/transactionReducer.ts (hunk 5 above)
Comment (juriczech) [reply suite-common/wallet-transactions/src/transactionReducer.ts]: https://github.com/trezor/trezor-suite/pull/5981/commits/86973889f127ee510e9e5b3f9fd0d10ad515d1c8

File: suite-common/wallet-transactions/src/transactionThunks.ts (hunk 4 above)
Comment (juriczech) [reply suite-common/wallet-transactions/src/transactionThunks.ts]: https://github.com/trezor/trezor-suite/pull/5981/commits/86973889f127ee510e9e5b3f9fd0d10ad515d1c8

File: suite-common/wallet-transactions/src/transactionThunks.ts (hunk 3 above)
Comment (juriczech) [reply suite-common/wallet-transactions/src/transactionThunks.ts]: Fixed

**Comment by juriczech** [comment]
Resolve conflicts after #6039 gets merged to develop. There are some changes to `replaceTransactionThunk` logic.

File: packages/suite/src/actions/wallet/transactionActions.ts [hunk 11]
```diff
@@ -1,198 +1,10 @@
...
-            dispatch({
-                type: TRANSACTION.FETCH_ERROR,
-                error: result ? result.payload.error : 'unknown error',
-            });
-        }
-    };
 
+// Note: This was not moved to suite-common due browser API that wouldn't work on mobile right now
+// (rest of the actions will be found in suite-common/wallet-core/transactions)
 export const exportTransactions =
```
Comment (Nodonisko) [inline packages/suite/src/actions/wallet/transactionActions.ts]: You can move it too. Just pass export api to it using extra. Should be super simple.

File: packages/suite/src/actions/wallet/__tests__/transactionActions.test.ts [hunk 12]
```diff
@@ -1,70 +1,57 @@
-import { configureStore } from '@suite/support/tests/configureStore';
-
-import transactionReducer from '@wallet-reducers/transactionReducer';
+import { transactionsReducer } from '@wallet-reducers';
```
Comment (Nodonisko) [inline packages/suite/src/actions/wallet/__tests__/transactionActions.test.ts:1]: This file should be moved to package too

File: packages/suite/src/middlewares/wallet/fiatRatesMiddleware.ts [hunk 13]
```diff
@@ -64,12 +64,17 @@ const fiatRatesMiddleware =
             case BLOCKCHAIN.FIAT_RATES_UPDATE:
                 api.dispatch(fiatRatesActions.onUpdateRate(action.payload));
                 break;
-            case TRANSACTION.ADD: {
-                // fetch historical rates for each added transaction
-                const { account, transactions } = action.payload;
-                api.dispatch(fiatRatesActions.updateTxsRates(account, transactions));
+            case transactionActions.addTransaction.type:
+                if (transactionActions.addTransaction.match(action))
```
Comment (Nodonisko) [inline packages/suite/src/middlewares/wallet/fiatRatesMiddleware.ts]: Place before switch, only place where we have if's inside of switch case is in legacy reducers.

File: packages/suite/src/middlewares/wallet/graphMiddleware.ts [hunk 14]
```diff
@@ -30,30 +30,32 @@ const graphMiddleware =
...
-                const discovery = api.dispatch(getDiscoveryForDevice());
-                if (
-                    discovery?.status === DISCOVERY.STATUS.COMPLETED &&
-                    transactions.some(t => (t.blockHeight ?? 0) > 0)
-                ) {
-                    api.dispatch(
-                        graphActions.updateGraphData([account], { newAccountsOnly: false }),
-                    );
+            case transactionActions.addTransaction.type:
+                if (transactionActions.addTransaction.match(action)) {
```
Comment (Nodonisko) [inline packages/suite/src/middlewares/wallet/graphMiddleware.ts]: Before switch

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts [hunk 15]
```diff
@@ -88,22 +87,30 @@ const storageMiddleware = (api: MiddlewareAPI<Dispatch, AppState>) => {
...
-                case TRANSACTION.REMOVE: {
-                    const { account } = action.payload;
-                    const device = accountUtils.findAccountDevice(account, api.getState().devices);
-                    // update only transactions for remembered device
-                    if (isDeviceRemembered(device)) {
-                        storageActions.removeAccountTransactions(account);
-                        api.dispatch(storageActions.saveAccountTransactions(account));
+                case transactionActions.addTransaction.type:
+                case transactionActions.removeTransaction.type:
+                    if (
```
Comment (Nodonisko) [inline packages/suite/src/middlewares/wallet/storageMiddleware.ts]: Before

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts [hunk 16]
```diff
@@ -88,22 +87,30 @@ const storageMiddleware = (api: MiddlewareAPI<Dispatch, AppState>) => {
...
+                        if (isDeviceRemembered(device)) {
+                            storageActions.removeAccountTransactions(account);
+                            api.dispatch(storageActions.saveAccountTransactions(account));
+                        }
                     }
                     break;
-                }
 
-                case TRANSACTION.RESET:
-                    storageActions.removeAccountTransactions(action.account);
```
Comment (Nodonisko) [inline packages/suite/src/middlewares/wallet/storageMiddleware.ts:104]: Before and use .match otherwise payload will be any

File: packages/suite/src/reducers/wallet/__tests__/transactionReducer.test.ts [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+            reducer(
+                { ...initialState, transactions },
+                {
+                    type: 'none',
+                },
+            ),
+        ).toEqual({ ...initialState, transactions });
+    });
+
+    it('TRANSACTION.RESET', () => {
```
Comment (Nodonisko) [inline packages/suite/src/reducers/wallet/__tests__/transactionReducer.test.ts]: Update names of tests

File: packages/suite/src/support/extraDependencies.ts [hunk 18]
```diff
@@ -20,13 +25,12 @@ export const extraDependencies: ExtraDependencies = {
...
     },
     actions: {
-        addTransaction: transactionActions.add,
-        removeTransaction: transactionActions.remove,
         setAccountLoadedMetadata: metadataActions.setAccountLoaded,
         setAccountAddMetadata: metadataActions.setAccountAdd,
     },
     actionTypes: {
         storageLoad: STORAGE.LOAD,
+        fiatRateUpdate: FIAT_RATES.TX_FIAT_RATE_UPDATE,
```
Comment (Nodonisko) [inline packages/suite/src/support/extraDependencies.ts]: You don't need to this through actionTypes and reducers, please convert fiatRateUpdate action to createAction syntax and then you will have typed payload for that case in common transactionReducer. Exactly same as @bouchja1 did for that two transaction actions before.

File: suite-common/wallet-core/src/transactions/transactionConstants.ts [hunk 19]
```diff
@@ -0,0 +1 @@
+export const modulePrefix = '@common/wallet-core/transactions';
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionConstants.ts]: Please scrap this file, we don't need files for one constant. Rename this constant to `actionPrefix` a move it to `transactionActions`, it belongs together anyway.

File: suite-common/wallet-core/src/transactions/transactionReducer.ts [hunk 20]
```diff
@@ -0,0 +1,145 @@
+import { accountsActions } from 'suite-common/wallet-core/src/index';
+
+import { Account, WalletAccountTransaction } from '@suite-common/wallet-types/libDev/src';
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionReducer.ts]: imports from `libDev/` 😬 that definitely should not happen. We need to ban this in eslint.

File: suite-common/wallet-core/src/transactions/transactionReducer.ts [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+import { settingsCommonConfig } from '@suite-common/suite-config/libDev/src';
+import { createReducerWithExtraDeps } from '@suite-common/redux-utils/libDev/src';
+
+import { transactionActions } from './transactionActions';
+
+export interface TransactionState {
+    isLoading: boolean;
+    error: string | null;
+    // Key is accountHash and value is sparse array of fetched txs
+    transactions: { [key: string]: WalletAccountTransaction[] };
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionReducer.ts]: please create `export type AccountHash = string` in suite-common/types/account and use it here instead

File: suite-common/wallet-core/src/transactions/transactionReducer.ts [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+    transactions: {},
+};
+
+interface TransactionRootState {
+    wallet: {
+        transactions: TransactionState;
+    };
+}
+
+const initializeAccount = (state: TransactionState, accountHash: string) => {
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionReducer.ts]: use type AccountHash instead of string

File: suite-common/wallet-core/src/transactions/transactionReducer.ts [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+import { transactionActions } from './transactionActions';
+
+export interface TransactionState {
+    isLoading: boolean;
+    error: string | null;
+    // Key is accountHash and value is sparse array of fetched txs
+    transactions: { [key: string]: WalletAccountTransaction[] };
+}
+
+export const initialState: TransactionState = {
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionReducer.ts]: `transactionsInitialState`

File: suite-common/wallet-core/src/transactions/transactionReducer.ts [hunk 24]
```diff
@@ -0,0 +1,145 @@
+import { accountsActions } from 'suite-common/wallet-core/src/index';
+
+import { Account, WalletAccountTransaction } from '@suite-common/wallet-types/libDev/src';
+import { findTransaction } from '@suite-common/wallet-utils/libDev/src';
+import { settingsCommonConfig } from '@suite-common/suite-config/libDev/src';
+import { createReducerWithExtraDeps } from '@suite-common/redux-utils/libDev/src';
+
+import { transactionActions } from './transactionActions';
+
+export interface TransactionState {
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionReducer.ts]: Transaction**s**State

File: suite-common/wallet-core/src/transactions/transactionReducer.ts [hunk 25]
```diff
@@ -0,0 +1,145 @@
+import { accountsActions } from 'suite-common/wallet-core/src/index';
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionReducer.ts]: please rename this file to transaction**s**Reducer.ts (you missing that "s" at the end of transaction)

File: suite-common/wallet-core/src/transactions/transactionActions.ts [hunk 26]
```diff
@@ -0,0 +1,60 @@
+import { createAction } from '@reduxjs/toolkit';
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionActions.ts]: please rename this file to transaction**s**Actions (you missing "s")

File: suite-common/wallet-core/src/transactions/transactionActions.ts [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+    } => ({
+        payload: {
+            transactions: transactions.map(t => enhanceTransaction(t, account)),
+            account,
+            page,
+        },
+    }),
+);
+
+export const transactionActions = {
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionActions.ts]: transaction"s"Actions

File: suite-common/wallet-core/src/transactions/transactionThunks.ts [hunk 28]
```diff
@@ -0,0 +1,232 @@
+import { accountsActions } from 'suite-common/wallet-core/src/index';
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionThunks.ts]: please rename this file to transaction"s"Thunks.ts

File: suite-common/wallet-core/src/transactions/transactionThunks.ts [hunk 29]
```diff
@@ -0,0 +1,232 @@
...
+
+import {
+    Account,
+    PrecomposedTransactionFinal,
+    TxFinalCardano,
+} from '@suite-common/wallet-types/libDev/src';
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionThunks.ts]: imports from libDev/ everywhere

File: suite-common/wallet-core/src/transactions/transactionActions.ts [hunk 30]
```diff
@@ -0,0 +1,60 @@
+import { createAction } from '@reduxjs/toolkit';
+
+import { Account, WalletAccountTransaction } from '@suite-common/wallet-types/libDev/src';
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionActions.ts]: imports from libDev everywhere

File: suite-common/wallet-core/src/transactions/transactionThunks.ts [hunk 31]
```diff
@@ -0,0 +1,232 @@
...
+                size: 0,
+                totalInput: '0',
+                totalOutput: '0',
+            },
+        };
+        dispatch(transactionActions.addTransaction({ transactions: [fakeTx], account }));
+    },
+);
+
+// TODO add this through extras and remove generator from transactionActions.ts
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionThunks.ts]: solved already?

File: suite-common/wallet-core/src/transactions/transactionReducer.ts [hunk 32]
```diff
@@ -0,0 +1,145 @@
...
+                extra.reducers.storageLoadTransactions,
+            )
+            .addMatcher(
+                action => action.type === extra.actionTypes.fiatRateUpdate,
+                extra.reducers.txFiatRateUpdate,
+            );
+    },
+);
+
+export const selectIsTransactionsLoading = (state: TransactionRootState) =>
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionReducer.ts]: It's good convetion to keep "is" and "Loading" together. => `selectIsLoadingTransactions`

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/wallet-core/src/transactions/transactionsThunks.ts [hunk 33]
```diff
@@ -49,7 +44,7 @@ export const replaceTransactionThunk = createThunk(
         // prepare replace actions for txs
         const actions = transactions.map(t => {
             const action = {
-                type: transactionActions.replaceTransaction.type,
+                type: transactionsActions.replaceTransaction.type,
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionsThunks.ts]: Use action creator

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/wallet-core/src/transactions/transactionsThunks.ts (hunk 33 above)
Comment (juriczech) [reply suite-common/wallet-core/src/transactions/transactionsThunks.ts]: solved

File: suite-common/test-utils/src/extraDependenciesMock.ts [hunk 34]
```diff
@@ -53,16 +53,23 @@ export const extraDependenciesMock: ExtraDependencies = {
...
     actionTypes: {
         storageLoad: mockActionType('storageLoad'),
     },
     reducers: {
         storageLoadBlockchain: mockReducer('storageLoadBlockchain'),
         storageLoadAccounts: mockReducer('storageLoadAccounts'),
+        storageLoadTransactions: mockReducer('storageLoadTransactions'),
+    },
+    utils: {
+        // TODO how to mock this
```
Comment (Nodonisko) [inline suite-common/test-utils/src/extraDependenciesMock.ts]: No need to TODO, this is actually mock. I would suggest just droppning sentence about phone and use just something generic like `Calling mocked "saveAs" to save ${data} into ....` because this mock could be also used in tests etc.

File: suite-common/wallet-types/src/transaction.ts [hunk 35]
```diff
@@ -220,3 +220,10 @@ export interface ReviewTransactionData {
         destinationTag?: string;
     };
 }
+
+export type TransactionFiatRateUpdateObject = {
```
Comment (Nodonisko) [inline suite-common/wallet-types/src/transaction.ts]: TransactionFiatRateUpdateObject => TransactionFiatRateUpdatePayload

**Review by Nodonisko — APPROVED** [review]

File: suite-common/test-utils/src/extraDependenciesMock.ts (hunk 34 above)
Comment (juriczech) [reply suite-common/test-utils/src/extraDependenciesMock.ts]: I missed the comment -> that was just note to myself and doesn't need to be there anymore. I will delete it.

**Comment by bosomt** [comment]
**QA OK**

see https://www.notion.so/satoshilabs/suite-native-transactions-storing-logic-5709-9cb815cef62647989be134af56667c30

Suite version: web 22.9.0 (667481022a3ebea89198c57e521f8cf8d7b78ae2)
